crates/katana/rpc/rpc/src/torii.rs

```diff
@@ -0,0 +1,173 @@
+use katana_provider::traits::transaction::TransactionProvider;
+use katana_rpc_api::torii::ToriiApiServer;
+use katana_rpc_types::error::torii::ToriiApiError;
+use katana_rpc_types::receipt::{MaybePendingTxReceipt, PendingTxReceipt};
+use katana_rpc_types::transaction::{TransactionsPage, TransactionsPageCursor};
+use katana_rpc_types_builder::ReceiptBuilder;
+use katana_tasks::TokioTaskSpawner;
+
+const MAX_PAGE_SIZE: usize = 100;
+
+#[derive(Clone)]
+pub struct ToriiApi {
+    sequencer: Arc<KatanaSequencer>,
+}
+
+impl ToriiApi {
+    pub fn new(sequencer: Arc<KatanaSequencer>) -> Self {
+        Self { sequencer }
+    }
+
+    async fn on_io_blocking_task<F, T>(&self, func: F) -> T
+    where
+        F: FnOnce(Self) -> T + Send + 'static,
+        T: Send + 'static,
+    {
+        let this = self.clone();
+        TokioTaskSpawner::new().unwrap().spawn_blocking(move || func(this)).await.unwrap()
+    }
+}
+
+#[async_trait]
+impl ToriiApiServer for ToriiApi {
+    async fn get_transactions(
+        &self,
+        cursor: TransactionsPageCursor,
+    ) -> RpcResult<TransactionsPage> {
+        self.on_io_blocking_task(move |this| {
+            let mut transactions = Vec::new();
+            let mut next_cursor = cursor.clone();
+
+            let provider = this.sequencer.backend.blockchain.provider();
+            let latest_block_number = this.sequencer.block_number().map_err(ToriiApiError::from)?;
+
+            if cursor.block_number > latest_block_number + 1 {
+                return Err(ToriiApiError::BlockNotFound.into());
+            }
+
+            if latest_block_number >= cursor.block_number {
+                for block_number in cursor.block_number..=latest_block_number {
+                    let mut block_transactions = provider
+                        .transactions_by_block(BlockHashOrNumber::Num(block_number))
+                        .map_err(ToriiApiError::from)?
+                        .ok_or(ToriiApiError::BlockNotFound)?;
+
+                    // If the block_number is the cursor block, slice the transactions from the txn
+                    // offset
+                    if block_number == cursor.block_number {
+                        block_transactions = block_transactions
+                            .into_iter()
+                            .skip(cursor.transaction_index as usize)
+                            .collect();
+                    }
+
+                    let block_transactions = block_transactions
+                        .into_iter()
+                        .map(|tx| {
+                            let receipt = ReceiptBuilder::new(tx.hash, provider)
+                                .build()
+                                .expect("Receipt should exist for tx")
+                                .expect("Receipt should exist for tx");
+                            (tx, MaybePendingTxReceipt::Receipt(receipt))
+                        })
+                        .collect::<Vec<_>>();
```

Comment:
Collecting all the transactions first before checking the limit and truncating it seems rather ineffecient. I think we should be able to pre compute how many transactions we need to fetch so as to be conversative with storage reads (and memory allocs), but `TransactionProvider` doesn't really provide methods for doing this easily. One way I could think of is by adding a method for getting the tx number of a tx and fetch them in a range up until the specified limit.

Not exactly a big issue for now, but something to be improved later.

Reply:
yeah this is a great optimization. i'll make a not of it. let's address it at some point